src/migration.py

```diff
@@ -0,0 +1,290 @@
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Module for transforming index table rows into local files."""
+
+import itertools
+import logging
+import typing
+from pathlib import Path
+
+from . import exceptions, types_
+from .discourse import Discourse
+from .docs_directory import calculate_table_path
+
+EMPTY_DIR_REASON = "<created due to empty directory>"
+GITKEEP_FILE = ".gitkeep"
+
+
+def _validate_row_levels(table_rows: list[types_.TableRow]):
+    """Check for invalid row levels.
+
+    Args:
+        table_rows: Table rows from the index file.
+
+    Raises:
+        InvalidRow exception if invalid row level is encountered.
+    """
+    level = 0
+    for i, row in enumerate(table_rows):
+        if row.level <= 0:
+            raise exceptions.InvalidTableRowError(f"Invalid level {row.level} in {row!=row.level}")
+        # Level increase of more than 1 is not possible.
+        if row.level > level and (difference := row.level - level) > 1:
+            raise exceptions.InvalidTableRowError(
+                f"Level difference of {difference} encountered in {row=!r}"
+            )
+        # Subdirectory but previous row is not a file.
+        if row.level > level and i > 0 and table_rows[i - 1].navlink.link:
+            raise exceptions.InvalidTableRowError(f"Invalid parent row for {row=!r}")
+
+        # Level decrease or same level is fine.
+        level = row.level
+
+
+def _migrate_gitkeep(gitkeep_meta: types_.GitkeepMeta, docs_path: Path):
+    """Write gitkeep file to docs directory.
+
+    Args:
+        gitkeep_meta: Gitkeep metadata from empty directory table row.
+        docs_path: Documentation folder path.
+
+    Returns:
+        Migration report for gitkeep file creation.
+    """
+    logging.info("migrate meta: %s", gitkeep_meta)
+
+    path = docs_path / gitkeep_meta.path
+    path.parent.mkdir(parents=True, exist_ok=True)
+    path.touch()
+    return types_.MigrationReport(
```

Comment:
Similar to other comment, do you think the `ActionReport` could be useful here? Maybe it would need to be generalised a bit?